src/Hooks.ts

```diff
@@ -143,3 +144,67 @@ export function useFocusState() {
 
   return focusState;
 }
+
+type EffectCallback = (() => void) | (() => () => void);
+
+// Inspired by same hook from react-navigation v5
+// See https://github.com/react-navigation/hooks/issues/39#issuecomment-534694135
+export const useFocusEffect = (callback: EffectCallback) => {
+  const navigation = useNavigation();
+
+  useEffect(() => {
+    let isFocused = false;
+    let cleanup: (() => void) | void;
+
+    if (navigation.isFocused()) {
+      cleanup = callback();
+      isFocused = true;
+    }
+
+    const focusSubscription = navigation.addListener('willFocus', () => {
+      // If callback was already called for focus, avoid calling it again
+      // The focus event may also fire on intial render, so we guard against runing the effect twice
+      if (isFocused) {
+        return;
+      }
+
+      cleanup && cleanup();
+      cleanup = callback();
+      isFocused = true;
+    });
+
+    const blurSubscription = navigation.addListener('willBlur', () => {
+      cleanup && cleanup();
+      cleanup = undefined;
+      isFocused = false;
+    });
+
+    return () => {
+      cleanup && cleanup();
+      focusSubscription.remove();
+      blurSubscription.remove();
+    };
+  }, [callback, navigation]);
```

Comment:
@slorber  If I run setParams in callback, navigation will change and invoke callback again, it will raise a infinite loop

Reply:
thanks for reporting this.

@satya164 how do you handle those loops in v5? I guess we should fix it the same way in v5 and here

Reply:
v5 doesn't update `navigation` prop on params update since `route` is a separate prop. regarding how to handle it in v4, user needs to add a check manually that params are out of date before trying to update them. they should do it anyways to prevent unnecessary updates to the screen regardless of the loop anyways since we don't do any equality checks like `setState` does.

Reply:
@satya164 I understand your point about preventing useless re-renders/infinite loops.

But if param value has to change, we still don't necessarily want to re-execute useFocusEffect again (at least user should be in control imho)

I mean:

This one should re-execute on userId param change

```js
useFocusEffect(useCallback(() => { ... },[params.userId]));
```
But this one should rather not re-execute even if any param changes:

```js
useFocusEffect(useCallback(() => { ... },[]));
```

So I think it's not correct to make this hook depend on navigation here, as the object is unstable it will tend to produce unwanted effect re-execution. I guess it will even re-execute if the parent re-renders for example.

I've opened another issue: https://github.com/react-navigation/hooks/issues/48